sysconf/sysconf.8
.\" Copyright (c) 2013-2016 Devin Teske
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FrauBSD: sysconf/sysconf.8 2016-01-09 13:08:21 -0800 freebsdfrau $
.\" $FreeBSD$
.\"
.Dd Nov 14, 2013
.Dt SYSCONF 8
.Os
.Sh NAME
.Nm sysconf
.Nd safely edit system configuration files
.Sh SYNOPSIS
.Nm
.Op Fl cDeFhinNqvx
.Op Fl f Ar file
.Op Fl R Ar dir
.Ar name Ns Op = Ns Ar value
.Ar ...
.Nm
.Op Fl cDeFhinNqvx
.Op Fl f Ar file
.Op Fl R Ar dir
.Fl a | A
.Sh DESCRIPTION
The
.Nm
utility retrieves
.Xr loader.conf 5
and
.Xr sysctl.conf 5
variables from the collection of system configuration files annd
allows processes with appropriate privilege to change values in
a safe and effictive manner.
.Pp
The following options are available:
.Bl -tag -width indent+
.It Fl a
Dump a list of all non-default configuration variables.
.It Fl A
Dump a list of all configuration variables
.Pq incl. defaults .
.It Fl c
Check if the value will change when assigning a new value.
If verbose
.Pq see Dq Fl v
prints a message stating whether a change would occur.
Exits with success if no change is necessary, else returns error status.
.It Fl D
Show default value(s) only (this is the same as setting SYS_CONFS to NULL or
passing `-f' with a NULL file-argument).
.It Fl e
Separate the name and the value of the variable(s) with
.Ql Li = .
This is useful for producing output which can be fed back to
.Xr sysctl 8 .
Ignored if
.Ql Fl n ,
.Ql Fl N ,
or
.Ql Fl F
is specified.
.It Fl f Ar file
Operate on the specified file(s) instead of the files obtained by reading the
.Sq loader_conf_files
entry in the
.Ev LOADER_DEFAULTS
file.
This option can be specified multiple times for additional files.
.It Fl F
Show only the last system configuration file each directive is in.
Either
.Xr loader.conf 5
or
.Xr sysctl.conf 5
file.
.It Fl h
Print a short usage message to stderr and exit.
.It Fl -help
Print a full usage statement to stderr and exit.
.It Fl i
Ignore unknown variables.
.It Fl n
Show only variable values, not their names.
.It Fl N
Show only variable names, not their values.
.It Fl q
Quiet.
Reverse previous occurrences of
.Fl v
flag.
.It Fl R Ar dir
Operate within the root directory
.Pq Sq Ar dir
rather than
.Pq Sq / .
.It Fl v
Verbose.
Print the pathname of the specific
.Xr rc.conf 5
file where the directive was found.
.It Fl -version
Print version information to stdout and exit.
.It Fl x
Remove variable(s) from specified file(s).
.El
.Pp
This utility works similar to
.Xr sysctl 8 .
It shares the `-e' and `-n' options
.Pq detailed above
and also has the same
.Ql name[=value]
syntax for querying/setting configuration options.
.Pp
However, while
.Xr sysctl 8
serves to query/modify MIBs in the entrant kernel,
.Nm
instead works on values in the system configuration files,
.Xr loader.conf 5
and
.Xr sysctl.conf 5 .
.Pp
The list of system configuration files for loader variables
.Pq CTLFLAG_TUN
is configured in the file
.Ql /boot/defaults/loader.conf
within the variable
.Ql loader_conf_files ,
which by-default contains a space-separated list of pathnames.
On all FreeBSD
systems, this defaults to "/boot/loader.conf /boot/loader.conf.local".
For runtime tunable sysctls, the list of system configuration files is
.Ql /etc/sysctl.conf
and
.Ql /etc/sysctl.conf.local .
.Pp
When supplied a variable name,
.Nm
will return the value of the variable.
If the variable does not appear in any
of the configured
.Ql loader_conf_files
or
sysctl runtime tunable configuration files,
an error is printed and error status is returned.
.Pp
When changing values of a given variable,
it does not matter if the variable appears in the system configuration files.
If the variable does not appear in any of the files, it is appended to either
the first pathname in
.Ql loader_conf_files
.Pq loader tunables
or
.Ql /etc/sysctl.conf .
Otherwise,
.Nm
will replace only the last-occurrence in the last-file found to contain the
variable.
This gets the value to take effect next boot without heavily
modifying these integral files
.Po
yet taking care to avoid file growth should
.Nm
be called repeatedly
.Pc .
.Sh ENVIRONMENT
The following environment variables are referenced by
.Nm :
.Bl -tag -width ".Ev LOADER_DEFAULTS"
.It Ev SYS_CONFS
Override default
.Ql loader_conf_files
.Pq even if NULL .
.It Ev LOADER_DEFAULTS
Location of
.Ql /boot/defaults/loader.conf
file.
.El
.Sh FILES
.Bl -tag -width ".Pa /etc/defaults/rc.conf" -compact
.It Pa /boot/defaults/loader.conf
.It Pa /boot/loader.conf
.It Pa /etc/sysctl.conf
.It Pa /etc/sysctl.conf.local
.El
.Sh EXAMPLES
Below are some simple examples of how
.Nm
can be used to query certain values from either
.Xr loader.conf 5
or
.Xr sysctl.conf 5
collection of system configuration files:
.Pp
.Nm
kernel
.Dl returns the value of kernel, usually /boot/kernel/kernel.
.Pp
.Nm
net.inet.ip.forwarding
.Dl returns 1 if IP forwarding between interfaces is enabled, otherwise 0.
.Pp
Working on other files, such as
.Xr crontab 5 :
.Pp
.Nm
-f /etc/crontab MAILTO
.Dl returns the value of the MAILTO setting Pq if configured .
.Pp
.Sh SEE ALSO
.Xr jail 8 ,
.Xr jexec 8 ,
.Xr jls 1 ,
.Xr loader 8 ,
.Xr loader.conf 5 ,
.Xr sysctl 8 ,
.Xr sysctl.conf 5
.\" .Sh HISTORY
.\" A
.\" .Nm
.\" utility first appeared in
.\" .Fx 10.1 .
.Sh AUTHORS
.An Devin Teske Aq dteske@FreeBSD.org


sysconf/sysconf.c
/*-
 * Copyright (c) 2013-2016 Devin Teske <dteske@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifdef __FrauBSD__
__FBSDID("$FrauBSD: sysconf/sysconf.c 2016-01-09 13:08:21 -0800 freebsdfrau $");
#else
__FBSDID("$FreeBSD$");
#endif

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "conf.h"
#include "sysconf.h"

/* Data to process */
unsigned int nargs = 0;

/* Extra display information */
char *pgm;			/* set to argv[0] by main() */
uint8_t show_equals = FALSE;	/* `-e' */

/* Function prototypes for private functions (see style(9)) */
void help(void);
void usage(void);

/*
 * Program description here.
 */
int
main(int argc, char *argv[])
{
	int ch;

	pgm = argv[0]; /* store a copy of invocation name */

	/*
	 * Process command-line options
	 */
	while ((ch = getopt(argc, argv, "-aAcDef:FhinNqR:vx")) != -1)
	{
		switch(ch) {
		case '-': /* long option */
			if (strcmp(argv[optind], "--help") == 0) help();
			else {
				warnx("illegal option -- %s", argv[optind]);
				usage();
			}
			break;
		case 'a': /* all */
			break;
		case 'A': /* all including defaults */
			break;
		case 'c': /* check only */
			break;
		case 'D': /* defaults only */
			break;
		case 'e': /* show equals */
			show_equals = TRUE;
			break;
		case 'f': /* file */
			break;
		case 'h': /* help/usage */
			usage();
			break;
		case 'i': /* ignore unknown */
			break;
		case 'n': /* value only */
			break;
		case 'N': /* name only */
			break;
		case 'q': /* quiet */
			break;
		case 'R': /* root dir */
			break;
		case 'v': /* verbose */
			break;
		case 'x': /* remove */
			break;
		case '?': /* unknown argument (based on optstring) */
		default: /* unhandled argument (based on switch) */
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	/* Process the contents of system configuration files */
	if (parse_sysconf_all() != 0 && errno != 0) exit(EXIT_FAILURE);

	/* Process remaining arguments as list of names to display */
	{
		int n = 0;
		char *cp;
		struct option *opt;

		for (cp = argv[0]; n < argc; cp = argv[n++])
		{
			nargs++;
			if ((opt = sysconf_get_option(cp)) == NULL) continue;
			printf(show_equals ? "%s=\"%s\"\n" : "%s: %s\n",
			       cp, opt->value);
		}
	}

	/* Display usage and exit if not given at least one name */
	if (nargs == 0) {
		warnx("no names provided");
		usage();
	}

	exit(EXIT_SUCCESS);
}

/*
 * Print short usage statement to stderr and exit with error status.
 */
void
usage(void)
{
	fprintf(stderr, "Usage: %s [OPTIONS] name[=value]\n", pgm);
	fprintf(stderr, "Try `%s --help' for more information.\n", pgm);
	exit(EXIT_FAILURE);
}

/*
 * Print long usage statement to stderr and exit with error status.
 */
void
help(void)
{
	fprintf(stderr, "Usage: %s [OPTIONS] name[=value]\n", pgm);
	fprintf(stderr, "Usage: %s [OPTIONS] -a | -A\n", pgm);
	fprintf(stderr, "OPTIONS:\n");
#define OPTFMT "\t%-9s %s\n"
	fprintf(stderr, OPTFMT, "-a",
	    "Dump a list of all non-default configuration variables.");
	fprintf(stderr, OPTFMT, "-A",
	    "Dump a list of all configuration variables (incl. defaults).");
	fprintf(stderr, OPTFMT, "-c",
	    "Check. Return success if no changes needed, else error.");
	fprintf(stderr, OPTFMT, "-D",
	    "Show default value(s) only (this is the same as setting");
	fprintf(stderr, OPTFMT, "",
	    "SYS_CONFS to NULL or passing `-f' with a NULL file-argument).");
	fprintf(stderr, OPTFMT, "-e",
	    "Separate the name and value of variable(s) with `='.");
	fprintf(stderr, OPTFMT, "-f file",
	    "Operate on the specified file(s) instead of normal collection");
	fprintf(stderr, OPTFMT, "",
	    "of files. Specify multiple times for additional files.");
	fprintf(stderr, OPTFMT, "-F",
	    "Show only the last configuration file each name appears in.");
	fprintf(stderr, OPTFMT, "-h",
	    "Print a short usage statement to stderr and exit.");
	fprintf(stderr, OPTFMT, "--help",
	    "Print this message to stderr and exit.");
	fprintf(stderr, OPTFMT, "-i",
	    "Ignore unknown names.");
	fprintf(stderr, OPTFMT, "-n",
	    "Show only variable values, not their names.");
	fprintf(stderr, OPTFMT, "-N",
	    "Show only variable names, not their values.");
	fprintf(stderr, OPTFMT, "-q",
	    "Quiet. Reverse previous occurrences of `-v' flag.");
	fprintf(stderr, OPTFMT, "-R dir",
	    "Operate within the root directory `dir' rather than `/'.");
	fprintf(stderr, OPTFMT, "-v",
	    "Verbose. Print pathname of the specific configuration file");
	fprintf(stderr, OPTFMT, "",
	    "where the name was found.");
	fprintf(stderr, OPTFMT, "--version",
	    "Print version information to stdout and exit.");
	fprintf(stderr, OPTFMT, "-x",
	    "Remove name(s) from specified file(s).");
	fprintf(stderr, "\n");
	fprintf(stderr, "ENVIRONMENT:\n");
#define ENVFMT "\t%-15s %s\n"
	fprintf(stderr, ENVFMT, "SYS_CONFS",
	    "Override default configuration files (NULL allowed).");
	fprintf(stderr, ENVFMT, "LOADER_DEFAULTS",
	    "Location of `/boot/defaults/loader.conf' file.");
	exit(EXIT_FAILURE);
}
